src/v/utils/hybrid_clock.h

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "base/likely.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/manual_clock.hh>
+
+#include <chrono>
+
+/// The hybrid_clock is a thin wrapper around the seastar::lowres_clock
+/// that can be swapped from seastar::manual_clock in tests by flipping
+/// a runtime flag (per shard). The timestamps and durations are using
+/// last bit as a flag indicating the type of clock used to produce
+/// them.
+class hybrid_clock {
+    struct duration_with_flag {
+        uint64_t nanoseconds{0};
+        bool manual{false};
+    };
+
+public:
+    // Same as std::chrono::steady_clock::rep
+    //
+    using duration = std::chrono::nanoseconds;
+    using rep = duration::rep;
+    using period = duration::period;
+    using time_point = std::chrono::time_point<hybrid_clock, duration>;
+    static constexpr bool is_steady = true;
+
+    inline static time_point now() noexcept;
+
+    inline static void update() noexcept;
+
+    inline static void advance(duration d);
+
```

Comment:
It's somewhat a naive suggestion so I imagine there's something obvious I'm missing, but if all we care about is being able to advance the clock in tests, could we get similar behavior if we declared some

```
class lowres_adjustable_clock {
    time_point now() {
        return ss::lowres_clock::now() + dur_added_;
    }
    void advance(duration d) {
        dur_added_ += d;
    }
    // ... all other methods fall through to ss::lowres_clock
private:
    std::chrono::nanoseconds dur_added_;
};
```

but without the caveats of bit hacks and off-by-one

Reply:
In this case the clock may drift because lowres_clock will be incremented. But I just realized that my version is also not going to cut it. In seastar functions like `sleep` or `get_units` are templates with parametrized clock. They treat manual_clock specially so you can actually do things like trigger a timeout. And with my implementation it will not happen unless you will use proper template specialization so it's kind of defeats the purpose of this.

Reply:
I guess that having templatized Clock is the only way.